doc/designs/quic-design/quic-api.md

```diff
@@ -1406,3 +1482,119 @@ calls.
+RFCs. A close reading of RFC 9000 suggests that the answer is, in principle,
+yes; however the RFC also grants explicit permission to make design choices in
+implementations which would preclude this:
+
+>An implementation might choose to defer allocating a stream ID to a stream until
+>it sends the first STREAM frame and enters this state, which can allow for
+>better stream prioritization.
+
+If an ID has not been allocated to a stream, obviously incoming data cannot be
+addressed to it. However, supposing that an implementation does not do this,
+RFC 9000 seems basically clear that it is valid for an application to create a
+stream locally, then receive data on it before sending anything:
+
+>The sending part of a stream that the endpoint initiates (types 0 and 2 for
+>clients, 1 and 3 for servers) is opened by the application. The "Ready"
+>state represents a newly created stream that is able to accept data from the
+>application.
+>
+>[...]
+>
+>For a bidirectional stream, the receiving part enters the "Recv" state when
+>the sending part initiated by the endpoint (type 0 for a client, type 1 for
+>a server) enters the "Ready" state.
+
+A peer is not generally notified of the creation of a stream which has not sent
+any data yet, since the creation of a stream is signalled only implicitly via
+the transmission of data in `STREAM` frames. However, a zero-length STREAM frame
+could presumably be used to effect such a notification. RFC 9000 contains no
+specific discussion of this possibility but does not preclude it. As such, in
+order to receive data on a locally-initiated bidirectional stream before sending
+any data on that stream, it would be necessary to either
+
+- Use a QUIC implementation which signals a bidirectional stream which has
+  not yet sent any data via a zero-length stream frame, or
+
+- Use an application protocol which can inform the peer of the stream ID
+  of the created stream in some application protocol-specific way.
+  This is somewhat less plausible because it would require an API between
+  the application and its QUIC library to inform the QUIC library
+  that the peer has in fact created a stream with a given ID and to
+  take its word for it. This is unlikely to be commonly available, especially as
+  application errors in usage of such an API would lead to internal
+  inconsistencies in QUIC connection state.
+
+Of course this discussion is somewhat esoteric as it is unclear why an
+application would want to create a locally-initiated stream and then have the
+peer transmit on it first, rather than simply use a remotely-initiated stream.
+Thus this discussion of this edge case is more of a curiosity, however for
+completeness it needs to be thought about in the API design.
+
+**Q. How should single-stream operation support locally and remotely-initiated
+streams?**
+
+Note that the ID of a stream depends on whether it is bidirectional and whether
```

Comment:
I am concerned that we're taking the single stream operation too far. IMO the single stream operation should be reserved for the case of client where the single stream is the locally initiated BIDI stream with ID 0 (and perhaps server where again the same stream (i.e. in case of server the remotely initiated BIDI stream with ID 0) is operated as the default stream of the connection. Anything else should use the multistream API, there should not be any default stream attach operation at all. I assume for multistream apps to be able to handle all stream SSL objects uniformly, it makes sense to provide the detach operation or even better maybe not and just provide an API to not create the default stream and require SSL_new_stream or SSL_accept_stream to be called for all streams.

Reply:
+1 on not having attach as per [earlier comment](https://github.com/openssl/openssl/pull/19769#discussion_r1159438797).

I'd also support not having detach -- once a steam gets the default, all multi-stream ops could disappear.  The default steam was intended to ease conversion of existing TLS based applications to use QUIC and not as functionality for QUIC aware applications.

Reply:
Even if we were only supporting a single stream we still have to support both locally and remotely initiated streams. Application protocols where the server sends first are entirely common, as are application protocols where the client sends first.

Reply:
That's an interesting consideration. Should we automatically attach the first remotely initiated bidi stream as the default stream if the application calls SSL_read() first? I.e., SSL_write() is first -> the default stream is the locally initiated one (0x00 on client, 0x01 on server) , SSL_read() first -> the default stream is the first remotely initiated one (0x01 on client, 0x00 on server).

Reply:
More generally, worth noting that there's a difference between complexity for the average application and the whole API we offer. The entire libssl API is enormous and no application uses all of it; most applications won't need all but a small subset of any API we define. So it's not too surprising we end up with some relatively 'niche' APIs.